app/js/ProjectPersistenceManager.js

```diff
@@ -61,26 +63,39 @@ module.exports = ProjectPersistenceManager = {
   },
 
   refreshExpiryTimeout: callbackify(refreshExpiryTimeout),
-  markProjectAsJustAccessed(project_id, callback) {
-    if (callback == null) {
-      callback = function (error) {}
-    }
-    const timer = new Metrics.Timer('db-bump-last-accessed')
-    const job = (cb) =>
-      db.Project.findOrCreate({ where: { project_id } })
-        .spread((project, created) =>
-          project
-            .update({ lastAccessed: new Date() })
-            .then(() => cb())
-            .error(cb)
+
+  init() {
+    fs.readdir(Settings.path.compilesDir, (err, projectIds) => {
+      if (err) {
+        logger.warn({ err }, 'cannot get project listing')
+        return
+      }
+
+      async.eachLimit(projectIds, 50, (projectId, cb) => {
+        const outputPdf = Path.join(
+          Settings.path.compilesDir,
+          projectId,
+          'output.pdf'
         )
-        .error(cb)
-    dbQueue.queue.push(job, (error) => {
-      timer.done()
-      callback(error)
+        fs.stat(outputPdf, (err, stats) => {
+          if (err) {
+            // Schedule for immediate cleanup
+            LAST_ACCESS.set(projectId, 0)
+          } else {
+            // Cleanup eventually.
+            LAST_ACCESS.set(projectId, stats.mtime.getTime())
+          }
+          cb()
+        })
+      })
     })
   },
 
+  markProjectAsJustAccessed(project_id, callback) {
+    LAST_ACCESS.set(project_id, Date.now())
```

Comment:
There's a slight inconsistency here, we mark the project by its `project_id` but when we read the directory we get `project_id-user_id`, so we're working with different keys.

```
clsi_1                    | LAST_ACCESS Map {
clsi_1                    |   '60c7195656b4630038c5f289-5dca84e11e71ae002ff73bd4' => 1624283908700,
clsi_1                    |   '60c7195656b4630038c5f289' => 1624284081074
clsi_1                    | }
```
Previously the clearing worked with the project_id only.  Not sure which is best here... we do clear the filestore cache when we expire the project, which argues in favour of using the project_id as the key, but we don't necessarily want to keep the compile directories themselves on disk for each user.

Reply:
Good spot. Fixed via e9c24deb9bedbef725a448d2900f8e9862ed7976

```
clsi_1                    | Map {
clsi_1                    |   '60b0dd39c418bc00598a0d22' => 1624545963805,
clsi_1                    |   '60950df909a951003b720945' => 1624622091312
clsi_1                    | }
```

The compile directories are cleared in a different process via `CompileManager.clearExpiredProjects`. But it looks like the output dirs are never cleared?

Reply:
> The compile directories are cleared in a different process via `CompileManager.clearExpiredProjects`. But it looks like the output dirs are never cleared?

That does seem to be the case, they used to be part of the compile directory so they were removed automatically when the compile directory was deleted.  Looks like I didn't carry that over when I moved them to a separate location.  I'll make a separate PR to add that back in.